src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,126 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        checkCapacity(size + 1);
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+        } else {
+            checkRange(index);
+            checkCapacity(size + 1);
+            T[] newElementData = generateNewArray(elementData.length);
+            System.arraycopy(elementData, 0, newElementData, 0, (size - 1) - index);
+            newElementData[index] = value;
+            System.arraycopy(elementData, index, newElementData, index + 1, size - index);
+            elementData = newElementData;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        checkCapacity(size + list.size());
+        T[] arrayFromList = (T[]) new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            arrayFromList[i] = list.get(i);
+        }
+        System.arraycopy(arrayFromList, 0, elementData, size, arrayFromList.length);
+        size = size + list.size();
+    }
+
+    private void grow() {
+        int oldCapacity = elementData.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        T[] newElementData = generateNewArray(newCapacity);
+        System.arraycopy(elementData, 0, newElementData, 0, size);
+        elementData = newElementData;
+    }
+
+    private T[] generateNewArray(int newCapacity) {
+        return (T[]) new Object[newCapacity];
+    }
+
+    private void checkCapacity(int capacity) {
```

Comment:
```suggestion
    private void checkCapacityAndGrow(int capacity) {
```
make it more informative